internal/grpc/services/publicshareprovider/publicshareprovider.go

```diff
@@ -139,29 +183,77 @@ func (s *service) CreatePublicShare(ctx context.Context, req *link.CreatePublicS
 	log := appctx.GetLogger(ctx)
 	log.Info().Str("publicshareprovider", "create").Msg("create public share")
 
+	// check that user has share permissions
+	// TODO Check if we need to stat again, because the client could send fake permissions
+	if !req.GetResourceInfo().GetPermissionSet().AddGrant {
+		err := errors.New("no share permission")
+		return &link.CreatePublicShareResponse{
+			Status: status.NewPermissionDenied(ctx, err, err.Error()),
+		}, nil
+	}
+
 	if !conversions.SufficientCS3Permissions(req.GetResourceInfo().GetPermissionSet(), req.GetGrant().GetPermissions().GetPermissions()) {
 		return &link.CreatePublicShareResponse{
-			Status: status.NewInvalid(ctx, "insufficient permissions to create that kind of share"),
+			Status: status.NewFailedPrecondition(ctx, nil, "insufficient permissions to create that kind of share"),
 		}, nil
 	}
 
 	if !s.isPathAllowed(req.ResourceInfo.Path) {
 		return &link.CreatePublicShareResponse{
-			Status: status.NewInvalid(ctx, "share creation is not allowed for the specified path"),
+			Status: status.NewFailedPrecondition(ctx, nil, "share creation is not allowed for the specified path"),
+		}, nil
+	}
+
+	// check that this is a valid share
+	if req.GetResourceInfo().GetId().GetOpaqueId() == req.GetResourceInfo().GetId().GetSpaceId() &&
+		req.GetResourceInfo().GetSpace().GetSpaceType() == "personal" {
+		return &link.CreatePublicShareResponse{
+			Status: status.NewFailedPrecondition(ctx, nil, "cannot create link on space root"),
 		}, nil
 	}
 
+	// check if a quicklink should be created
+	quickString, ok := req.GetResourceInfo().GetArbitraryMetadata().GetMetadata()["quicklink"]
```

Comment:
The `GetMetadata()` can return nil. Will be good to check it.	
```go
if m := sres.GetInfo().GetArbitraryMetadata().GetMetadata(); m != nil {
	quickString := m["quicklink"]
}
```

Reply:
ok, thanks.